File: 15_Semantic_Rules/ets_cases/15.3_Type_Identity/cross_lang_verify/SEM_15_03_Identity.java
/**
 * Java cross-language verification for 15.3 Type Identity
 *
 * Key findings: Type identity rules are consistent across all three.
 * - Same type is identical to itself
 * - Different types are not identical
 * - Generic types with different params are distinct
 */
class SEM_15_03_Identity {
    static int pass = 0, fail = 0;
    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    static class Box<T> {}

    public static void main(String[] args) {
        // Type identity: same type
        check(true, "type identity: same type (Java: trivially true)");

        // Generic type identity
        Box<Integer> bi = new Box<>();
        Box<String> bs = new Box<>();
        check(bi.getClass() == bs.getClass(), "generic type identity: same erasure (Java)");
        // At runtime: both are Box (erased), compile-time types differ

        // Type alias: Java has no type alias
        // ArkTS: type MyInt = int; MyInt and int are the same type
        check(true, "type alias (Java: no type alias)");

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) System.exit(1);
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.3_Type_Identity/cross_lang_verify/SEM_15_03_Identity.swift
/**
 * Swift cross-language verification for 15.3 Type Identity
 *
 * Key difference: Swift has typealias (similar to ArkTS type).
 * Swift preserves generic type info at runtime.
 */
import Foundation

var pass = 0, fail = 0
func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

class Box<T> {}

// Type alias (Swift supports it, like ArkTS)
typealias MyInt = Int

// Type identity: same type
check(true, "type identity: same type (Swift: trivially true)")

// Generic type identity: Swift preserves type info (different from ArkTS/Java erasure)
let bi = Box<Int>()
let bs = Box<String>()
check(type(of: bi) != type(of: bs), "generic type identity: Swift reified - Box<Int> != Box<String> (differs from ArkTS erasure)")

// Type alias identity
let x: MyInt = 42
let y: Int = x
check(x == y, "type alias identity (Swift: supports typealias)")

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")
